### ISSUE TITLE
When BatchNormalization layer is shared Siamese-style, moving_variance is calculated incorrectly.

### ISSUE THREAD

**Issue by ozabluda** [keras-team/keras#8794]
When `BatchNormalization` layer is shared Siamese-style, `moving_variance` is calculated incorrectly. In the code below, if `BatchNormalization` is not shared (`if` branch of the conditional) the final weights are correct: 
```
[ 1.00000203  1.          0.99995679  0.99995679]
```
the order is
```
[gamma, beta, moving_mean, moving_variance]
```
but if the `BatchNormalization` layer is shared (`else` branch of the conditional), the final weights are incorrect:
```
[  0.00000000e+00   6.13518556e-37   9.99296010e-01   0.00000000e+00]
```
namely, `moving_variance` (last weight) is zero, which is the most likely, the root cause of it. It's as if one of the shared `BatchNormalization` layers, overwrites `moving_variance` of the other. Note that `moving_mean` is calculated correctly in a shared way. `gamma` and `beta` also work "correctly" to compensate for the wrong `moving_variance`.

```
# Script for reproducing a BatchNormalization bug
# https://gist.github.com/ozabluda/11326307bbf13936f18063a1a4165adc

from keras.models import Sequential, Model
from keras.layers import Dense, BatchNormalization, Input, concatenate
import numpy as np

m = Sequential([
    BatchNormalization(input_shape=(1,),
                       center=True,
                       scale=True,
                       gamma_initializer='zero',
                       beta_initializer='one',
                       moving_mean_initializer='zero',
                       moving_variance_initializer='zero',
                       epsilon=0.00001
    ),
])
m.summary()
print(m.layers[0].weights)

if False: # not shared
    x  = np.array([0, 2])
    y  = np.array([0, 2])

    m.compile(optimizer='sgd', loss='mse')
    print(m.evaluate(x,y), m.predict(x))
    print(np.array(m.layers[0].get_weights()).ravel())

    m.fit(x, y, verbose=0, epochs=1000)

    print(m.evaluate(x,y), m.predict(x))
    print(np.array(m.layers[0].get_weights()).ravel())

else: #shared
    input_a = Input(shape=(1,))
    input_b = Input(shape=(1,))

    processed_a = m(input_a)
    processed_b = m(input_b)

    c = concatenate([processed_a, processed_b])
    c = Dense(1, kernel_initializer='ones', use_bias=False, trainable=False)(c)
    s = Model([input_a, input_b], c)
    s.compile(optimizer='sgd', loss='mse')
    s.summary()

    x0 = np.array([0])
    x1 = np.array([2])
    x  = [x0,x1]
    y  = np.array([0])

    print(np.array(s.layers[2].layers[0].get_weights()).ravel())

    s.fit(x, y, verbose=0, epochs=200)

    print(s.evaluate(x,y), s.predict(x))
    print(np.array(s.layers[2].layers[0].get_weights()).ravel())
```

========

- [x] Check that you are up-to-date with the master branch of Keras. [OZ: Keras version is after https://github.com/keras-team/keras/pull/8785 

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. [OZ: TF is 1.4.1]

**Comment by ozabluda** [comment]
more experimentation:
```
    x0 = np.array([0, 4])
    x1 = np.array([2, 0])
    x  = [x0,x1]
    y  = np.array([0, 0])
```
also produces wrong `moving_variance`:
```
[  0.00000000e+00   6.13518556e-37   1.49986613e+00   2.49955583e+00]
```
but now my hypothesis is that BN calculates average of 2 variances:

```
    print(np.var(x), np.var(x0), np.var(x1))
```
outputs
```
(2.75, 4.0, 1.0)
```
and 
```
2.49955583e00 = (4.0+1.0)/2
```

**Comment by fchollet** [comment]
If you share a BN layer:

- across two independent models: when training model A, the BN layer's weights will be updated with the batch statistics of the data from model A; and reciprocally. Basically: we only run the updates that correspond to the model you're training
- inside a same model (what you are doing): all updates get run, thus they get averaged across all data flows (because [the corresponding updates are incremental](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/moving_averages.py#L88)). This is what you're seeing. This is in fact the correct behavior.

If you want to compute updates for one data flow only, you can call `layer(x, training=False)` for one of the branches. E.g.:

```python
x1 = Input(...)
x2 = Input(...)
layer = BatchNormalization(...)
y1 = layer(x1)
y2 = layer(x2, training=False)
```

**Comment by ozabluda** [comment]
That's definitely not a correct behavior, but I can't tell if it's a bug/misfeature in Keras or the TF backend (if so, I'll file the bug report there). 

If I understand correctly, shared BN independently calculates `moving_mean` and `moving_variance`, then averages them at some later point (after each minibatch?), which for the combined dataflow actually produces correct result for `moving_mean`, but a nonsensical one for `moving_variance` [1]. A correct behavior would be to first average `moving_mean`, and only then calculate and average `moving_variance` relative to this shared `moving_mean`.

[1] Although for individual dataflows, neither `moving_mean` nor `moving_variance` are correct, which I think is intuitive, since a single set of `moving_mean, moving_variance` can't possibly fit two different datasets.

**Comment by hoangcuong2011** [comment]
Just in case anyone who happened to face this issue:

I fixed the problem by writing a custom shared Batch normalization layer. That batch norm layer has its own beta and gamma parameters as usual. It, however, receives two inputs, and the mean and variance are computed based on statistics from both the two inputs. 

The beta and gamma vectors are interacted with shared mean and variance.

It is actually very interesting to write that custom layer.

**Comment by 0xfiras** [comment]
> It is actually very interesting to write that custom layer.

Gist please?

**Comment by MaxChu719** [comment]
When I have a large number of shared weighted subnetworks (~350), training will be normal but when performing inference, the result will be nan... I am really interested to see how you implement a shared weight version of batch normalization. Would you mind sharing your code?

**Comment by hoangcuong2011** [comment]
Hello @MaxChu719 
Sorry I wrote the code for a project at a company, and thus I am not able to share it publicly.

I also want o say that it is trick but not that super difficult to implement. What you need to do is to look at the original code of Batch Normalization (https://github.com/keras-team/keras/blob/master/keras/layers/normalization.py#L16), and then extend it. The inputs should have both sides (left and right) for Siamese networks as well.
